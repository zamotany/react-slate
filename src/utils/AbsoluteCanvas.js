/* @flow */

/* eslint-disable no-param-reassign */

import stripAnsi from 'strip-ansi';
import AnsiParser from 'ansi-parser';

type CanvasSize = {
  width: number,
  height: number,
};

type Char = {
  style: string,
  content: string,
};

function isEmpty(value: ?(string[])) {
  return !value || !value.length;
}

function mergeAnsiStrings(bottomString: string, topString: string) {
  const bottom: Char[] = AnsiParser.parse(bottomString);
  const top: Char[] = AnsiParser.parse(topString);
  const output: Char[] = [];

  const length = Math.max(bottom.length, top.length);
  for (let i = 0; i < length; i++) {
    const char =
      top[i] && top[i].content && top[i].content !== '\0'
        ? top[i]
        : bottom[i] || { style: '', content: '\0' };
    output.push(char);
  }

  return AnsiParser.stringify(output).replace(/\u001b\[0m$/, '');
}

export default class AbsoluteCanvas {
  positiveLayers = [];
  negativeLayers = [];

  noContentLine: string;
  size: CanvasSize;

  constructor({ width, height }: CanvasSize) {
    this.size = { width, height };
    this.noContentLine = '\0'.repeat(width);
  }

  atLayer(layerIndex: number) {
    if (!Number.isInteger(layerIndex)) {
      throw new Error('Layer index (z-index) must be an integer.');
    }

    if (layerIndex > -1) {
      this.positiveLayers[layerIndex] = isEmpty(this.positiveLayers[layerIndex])
        ? new Array(this.size.height).fill(this.noContentLine)
        : this.positiveLayers[layerIndex];
      return this.positiveLayers[layerIndex];
    }
    const index = -layerIndex;
    this.negativeLayers[index] = isEmpty(this.negativeLayers[index])
      ? new Array(this.size.height).fill(this.noContentLine)
      : this.negativeLayers[index];
    return this.negativeLayers[index];
  }

  appendTree(
    nestedTree: string[],
    { x, y, z }: { x: number, y: number, z: number }
  ) {
    const layer = this.atLayer(z);
    for (let i = 0; i < nestedTree.length && y + i < layer.length; i++) {
      layer[y + i] = mergeAnsiStrings(
        layer[y + i],
        `${'\0'.repeat(x)}${nestedTree[i]}`
      ).substr(0, this.size.width);
    }
  }

  merge(bottomCanvas: string[], topCanvas: string[]) {
    for (let lineIndex = 0; lineIndex < this.size.height; lineIndex++) {
      if (
        // If both strings have content.
        bottomCanvas[lineIndex] &&
        bottomCanvas[lineIndex] !== this.noContentLine &&
        topCanvas[lineIndex] &&
        topCanvas[lineIndex] !== this.noContentLine
      ) {
        bottomCanvas[lineIndex] = mergeAnsiStrings(
          bottomCanvas[lineIndex],
          topCanvas[lineIndex]
        );
      } else if (
        // If only top string has content.
        (!bottomCanvas[lineIndex] ||
          bottomCanvas[lineIndex] === this.noContentLine) &&
        topCanvas[lineIndex] &&
        topCanvas[lineIndex] !== this.noContentLine
      ) {
        bottomCanvas[lineIndex] = topCanvas[lineIndex];
      } else {
        bottomCanvas[lineIndex] = bottomCanvas[lineIndex] || this.noContentLine;
      }

      const lineLength = stripAnsi(bottomCanvas[lineIndex]).length;
      if (lineLength < this.noContentLine.length) {
        bottomCanvas[lineIndex] += '\0'.repeat(
          this.noContentLine.length - lineLength
        );
      }
    }
    return bottomCanvas;
  }

  flatten(relativeCanvas: string[]) {
    // eslint-disable-next-line no-unused-vars
    let flatCanvas = this.negativeLayers.reduce(
      (acc: string[], currentCanvas: string[]) =>
        this.merge(currentCanvas, acc),
      new Array(this.size.height).fill(this.noContentLine)
    );
    flatCanvas = this.merge(flatCanvas, relativeCanvas);
    return this.positiveLayers
      .reduce(this.merge.bind(this), flatCanvas)
      .map(line => line.replace(/\0/g, ' '));
  }
}
