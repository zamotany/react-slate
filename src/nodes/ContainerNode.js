/* @flow */

import readline from 'readline';
import enhanceConsole from '../effects/enhanceConsole';
import {
  hideCursor,
  clearOnExit,
  clearScrollBackOnExit,
} from '../effects/terminal';
import ChunkNode from './ChunkNode';
import { mergeCanvas, getCanvas } from '../utils/layout';

type Options = {
  debug: boolean,
  hideCursor: boolean,
  clearScreenOnExit: boolean,
  clearScrollbackOnExit: boolean,
  exitOnWarning: boolean,
  exitOnError: boolean,
};

export default class ContainerNode {
  children: ChunkNode[] = [];
  frontBuffer: string = '';
  backBuffer: string = '';
  stream: any = null;
  options: Options;
  canvasSize: { width: number, height: number };

  constructor(stream: any, opts?: Options) {
    this.options = {
      debug: false,
      exitOnError: false,
      exitOnWarning: false,
      hideCursor: false,
      clearScreenOnExit: false,
      clearScrollbackOnExit: false,
      ...(opts || {}),
    };
    this.stream = stream;

    // @TODO: handle resize
    this.canvasSize = {
      width: this.stream.columns,
      height: this.stream.rows - 1,
    };

    enhanceConsole({
      exitOnError: this.options.exitOnError,
      exitOnWarning: this.options.exitOnWarning,
    });

    if (this.options.hideCursor) {
      hideCursor(this.stream);
    }

    if (this.options.clearScreenOnExit) {
      clearOnExit(this.stream);
    }

    if (this.options.clearScrollbackOnExit) {
      clearScrollBackOnExit(this.stream);
    }
  }

  invalidateParent = () => {
    /* NOOP */
  };

  appendChild(child: ChunkNode) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    this.children.push(child);
  }

  prependChild(child: ChunkNode, childBefore: ChunkNode) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    const index = this.children.indexOf(childBefore);
    this.children.splice(index, 0, child);
  }

  removeChild(child: ChunkNode) {
    const index = this.children.indexOf(child);
    this.children.splice(index, 1);
  }

  flush() {
    // @TODO: this buffer/optimization/slitting logic needs to be refactored
    // @TODO: draw damage to screen only instead of everything
    const canvas = getCanvas(this.canvasSize);
    this.frontBuffer = mergeCanvas(
      canvas,
      this.children.reduce(
        (acc, child) => [...acc, ...child.render(canvas)],
        []
      )
    ).join('\n');

    readline.cursorTo(this.stream, 0, 0);
    readline.clearScreenDown(this.stream);
    this.stream.write(this.frontBuffer);
  }
}
