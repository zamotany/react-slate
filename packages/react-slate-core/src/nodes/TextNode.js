/* @flow */

import type { Props } from '../types';
import ContainerNode from './ContainerNode';
import ChunkNode from './ChunkNode';

/**
 * Simple node containing text as props.children.
 */
export default class TextNode {
  static componentName = 'TEXT_NODE';

  container: ContainerNode;
  props: Props;
  parent: ChunkNode;

  constructor(container: ContainerNode, props: Props) {
    this.container = container;
    this.props = props;
  }

  replaceChildren(children: string) {
    this.props.children = children;
  }
}
