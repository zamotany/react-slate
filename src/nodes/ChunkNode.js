/* @flow */

import type { Props, Element, Position } from '../types';
import TextNode from './TextNode';
import ContainerNode from './ContainerNode';

type ChunkNodePros = Props & {
  x: number,
  y: number,
  children: any,
};

export default class ChunkNode {
  static componentName = 'CHUNK_NODE';

  props: ChunkNodePros;
  container: ContainerNode;
  parent: ChunkNode | ContainerNode;
  children = [];
  parentsOffset: Position = { x: 0, y: 0 };
  previousPosition: Position;
  hasChildrenChanged: boolean = true;
  hasParentChanged: boolean = false;
  memoizedElements: Element[] = [];

  constructor(container: ContainerNode, props: ChunkNodePros) {
    this.container = container;
    this.props = props;
    this.previousPosition = { x: props.x, y: props.y };
  }

  hasPositionChanged() {
    return (
      this.props.x !== this.previousPosition.x ||
      this.props.y !== this.previousPosition.y
    );
  }

  setParentsOffset({ x, y }: Position) {
    this.parentsOffset.x = x;
    this.parentsOffset.y = y;
  }

  getChildOffset() {
    return {
      x: this.props.x + this.parentsOffset.x,
      y: this.props.y + this.parentsOffset.y,
    };
  }

  invalidateParent() {
    // Invalidate the whole path from this node up to the top.
    this.hasChildrenChanged = true;
    this.parent.invalidateParent();
  }

  appendChild(child: ChunkNode | TextNode) {
    this.invalidateParent();
    // eslint-disable-next-line no-param-reassign
    this.appendInitialChild(child);
  }

  appendInitialChild(child: ChunkNode | TextNode) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    if (child instanceof ChunkNode) {
      child.setParentsOffset(this.getChildOffset());
    }
    this.children.push(child);
  }

  removeChild(child: ChunkNode | TextNode) {
    this.invalidateParent();
    const index = this.children.indexOf(child);
    this.children.splice(index, 1);
  }

  render() {
    // When parent changes (eg. positions is updated) it might affect children.
    const hasParentChanged = this.hasParentChanged || this.hasPositionChanged();
    if (
      this.hasChildrenChanged ||
      !this.memoizedElements.length ||
      hasParentChanged
    ) {
      // At least one of the children has changed or position of the parent
      // has changed.

      this.memoizedElements = [];
      this.hasChildrenChanged = false;

      for (const child of this.children) {
        if (child instanceof ChunkNode) {
          // Child is a ChunkNode, which means it's a nested subtree. In this case, move
          // responsibility of appending elements to the container to this subtree and
          // memoize the returned elements.

          // Parents offset need to be corrected, otherwise this subtree would have
          // the old offset from previous render.
          child.hasParentChanged = hasParentChanged;
          child.setParentsOffset(this.getChildOffset());
          this.memoizedElements.push(...child.render());
        } else {
          // Child is a TextNode, so we can just create element object and append it to
          // memoized elements and to the container.
          const parentsOffset = this.getChildOffset();
          const element = {
            x: 0,
            y: 0,
            parentsOffsetX: parentsOffset.x,
            parentsOffsetY: parentsOffset.y,
            text: child.props.children,
          };
          this.memoizedElements.push(element);
          this.container.appendElement(element);
        }
      }
    } else {
      this.memoizedElements.forEach(element => {
        this.container.appendElement(element);
      });
    }

    // Return memoized elements (old or freshly created new one), so that the parent can memoize
    // it's subtree.
    return this.memoizedElements;
  }
}
