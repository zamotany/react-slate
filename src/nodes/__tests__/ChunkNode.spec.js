/* @flow */

import ChunkNode from '../ChunkNode';
import TextNode from '../TextNode';
import ContainerNode from '../ContainerNode';
import AbsoluteCanvas from '../../utils/AbsoluteCanvas';
import colors from '../../constants/colors';

function withChildren(instance, children) {
  children.forEach(instance.appendInitialChild.bind(instance));
  return instance;
}

function getNodeProps(props = {}) {
  return {
    marginTop: 0,
    marginBottom: 0,
    marginRight: 0,
    marginLeft: 0,
    paddingTop: 0,
    paddingBottom: 0,
    paddingRight: 0,
    paddingLeft: 0,
    width: -1,
    height: 0,
    children: null,
    inline: false,
    fixed: false,
    x: 0,
    y: 0,
    z: 0,
    stylizeArgs: {},
    ...props,
  };
}

function getCanvasMock() {
  return new AbsoluteCanvas({ width: 10, height: 10 });
}

describe('nodes/ChunkNode', () => {
  describe('render method', () => {
    it('should render TextNodes one after another', () => {
      // <Text>
      //   {'Text1'}
      //   {'Text2'}
      // </Text>
      const container = (({}: any): ContainerNode);
      // $FlowFixMe
      const rootNode = new ChunkNode(container, getNodeProps());
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1' })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text2' })
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual(['Text1Text2']);
    });

    it('should render TextNodes separated by 3rd node with \\n', () => {
      // <Text marginTop={1}>
      //   {'Text1'}
      //   {'\n'}
      //   {'Text2'}
      // </Text>
      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          marginTop: 1,
        })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1' })
      );
      rootNode.appendInitialChild(new TextNode(container, { children: '\n' }));
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text2' })
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual([
        '',
        'Text1',
        'Text2',
      ]);
    });

    it('should render TextNodes separated by \\n', () => {
      // <Text marginTop={1}>
      //   {'Text1\n'}
      //   {'Text2'}
      // </Text>
      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(container, getNodeProps({ marginTop: 1 }));
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1\n' })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text2' })
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual([
        '',
        'Text1',
        'Text2',
      ]);
    });

    it('should render ChunkNodes one below another', () => {
      // <Text>
      //   <Text>Text1</Text>
      //   <Text>Text2</Text>
      // </Text>
      const container = (({}: any): ContainerNode);
      // $FlowFixMe
      const rootNode = new ChunkNode(container, getNodeProps());
      rootNode.appendInitialChild(
        withChildren(new ChunkNode(container, getNodeProps()), [
          new TextNode(container, { children: 'Text1' }),
        ])
      );
      rootNode.appendInitialChild(
        withChildren(new ChunkNode(container, getNodeProps()), [
          new TextNode(container, { children: 'Text2' }),
        ])
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual([
        'Text1',
        'Text2',
        '',
      ]);
    });

    it('should render TextNode and inlined ChunkNodes in the same line', () => {
      // <Text>
      //   Text1
      //   <Text>Text2</Text>
      // </Text>
      const container = (({}: any): ContainerNode);
      // $FlowFixMe
      const rootNode = new ChunkNode(container, getNodeProps());
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1' })
      );
      rootNode.appendInitialChild(
        withChildren(new ChunkNode(container, getNodeProps({ inline: true })), [
          new TextNode(container, { children: 'Text2' }),
        ])
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual(['Text1Text2']);
    });

    it('should add margins (shallow)', () => {
      // <Text
      //   marginTop={2}
      //   marginBottom={1}
      //   marginLeft={2}
      //   marginRight={2}
      // >
      //   {'Text1\n'}
      //   {'Text2'}
      // </Text>
      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          marginTop: 2,
          marginBottom: 1,
          marginLeft: 2,
          marginRight: 2,
        })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1\n' })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text2' })
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual([
        '',
        '',
        '\0\0Text1\0\0',
        '\0\0Text2\0\0',
        '',
      ]);
    });

    it('should add margins indentation (nested)', () => {
      // <Text
      //   marginTop={2}
      //   marginBottom={1}
      //   marginLeft={2}
      //   marginRight={2}
      // >
      //   Text1
      //   <Text
      //     marginTop={1}
      //     marginBottom={1}
      //     marginLeft={1}
      //     marginRight={1}
      //   >
      //     Text2
      //   </Text>
      //   Text3
      // </Text>
      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          marginTop: 2,
          marginBottom: 1,
          marginLeft: 2,
          marginRight: 2,
        })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1' })
      );
      rootNode.appendInitialChild(
        withChildren(
          new ChunkNode(
            container,
            getNodeProps({
              marginTop: 1,
              marginRight: 1,
              marginBottom: 1,
              marginLeft: 1,
            })
          ),
          [new TextNode(container, { children: 'Text2' })]
        )
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text3' })
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual([
        '',
        '',
        '\0\0Text1\0\0',
        '\0\0\0Text2\0\0\0',
        '\0\0\0\0',
        '\0\0Text3\0\0',
        '',
      ]);
    });

    it('should render a valid representation of the tree', () => {
      // <Text
      //   marginTop={1}
      //   marginBottom={1}
      //   paddingLeft={2}
      //   paddingRight={2}
      //   paddingTop={1}
      //   paddingBottom={1}
      //   width={9}
      //   height={5}
      // >
      //   <Text marginLeft={1} marginRight={1}>
      //     <Text inline>A</Text>
      //     <Text inline>B</Text>
      //     <Text inline>C</Text>
      //   </Text>
      //   <Text marginTop={1}>
      //     Text1
      //   </Text>
      // </Text>

      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          marginTop: 1,
          marginBottom: 1,
          paddingLeft: 2,
          paddingRight: 2,
          paddingTop: 1,
          paddingBottom: 1,
          width: 9,
          height: 5,
        })
      );
      rootNode.appendInitialChild(
        withChildren(
          new ChunkNode(
            container,
            getNodeProps({
              marginLeft: 1,
              marginRight: 1,
            })
          ),
          [
            withChildren(
              new ChunkNode(container, getNodeProps({ inline: true })),
              [new TextNode(container, { children: 'A' })]
            ),
            withChildren(
              new ChunkNode(container, getNodeProps({ inline: true })),
              [new TextNode(container, { children: 'B' })]
            ),
            withChildren(
              new ChunkNode(container, getNodeProps({ inline: true })),
              [new TextNode(container, { children: 'C' })]
            ),
          ]
        )
      );
      rootNode.appendInitialChild(
        withChildren(
          new ChunkNode(
            container,
            getNodeProps({
              marginTop: 1,
            })
          ),
          [new TextNode(container, { children: 'Text1' })]
        )
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual([
        '\0'.repeat(9),
        ' '.repeat(9),
        '  \0ABC\0  ',
        `${' '.repeat(4)}${'\0'.repeat(5)}`,
        '  Text1  ',
        `${' '.repeat(4)}${'\0'.repeat(5)}`,
        '\0'.repeat(9),
      ]);
    });

    it('should append fixed subtree in global canvas', () => {
      /**
       * <Text style={{ position: 'fixed', left: 1, top: 1 }}>
       *   <Text style={{ marginBottom: 1 }}>Text1</Text>
       *   {'Text2A\nText2B'}
       * </Text>
       */

      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          fixed: true,
          x: 1,
          y: 1,
        })
      );
      rootNode.appendInitialChild(
        withChildren(
          new ChunkNode(
            container,
            getNodeProps({
              marginBottom: 1,
            })
          ),
          [new TextNode(container, { children: 'Text1' })]
        )
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text2A\nText2B' })
      );

      const absoluteCanvas = getCanvasMock();
      expect(absoluteCanvas.flatten(rootNode.render(absoluteCanvas))).toEqual([
        ' '.repeat(10),
        ' Text1    ',
        ' '.repeat(10),
        ' Text2A   ',
        ' Text2B   ',
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
      ]);
    });

    it('should properly layout nested fixed subtree', () => {
      /**
       * <Text style={{ position: 'fixed', left: 1, top: 1 }}>
       *   {'Text1'}
       *   <Text style={{
       *     position: 'fixed',
       *     left: 2,
       *     top: 0
       *   }}>Text2</Text>
       * </Text>
       */

      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          fixed: true,
          x: 1,
          y: 1,
        })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1' })
      );
      rootNode.appendInitialChild(
        withChildren(
          new ChunkNode(
            container,
            getNodeProps({
              fixed: true,
              x: 2,
              y: 0,
            })
          ),
          [new TextNode(container, { children: 'Text2' })]
        )
      );

      const absoluteCanvas = getCanvasMock();
      expect(absoluteCanvas.flatten(rootNode.render(absoluteCanvas))).toEqual([
        '  Text2   ',
        ' Text1    ',
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
        ' '.repeat(10),
      ]);
    });

    it('should use custom render prop', () => {
      const customRender = jest.fn((instance, localCanvas, absoluteCanvas) => {
        expect(instance.children.length).toBe(1);
        expect(absoluteCanvas).toBeDefined();
        localCanvas.canvas.push('CustomText');
        return localCanvas;
      });
      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          fixed: true,
          x: 1,
          y: 1,
          render: customRender,
        })
      );
      rootNode.appendInitialChild(
        new TextNode(container, { children: 'Text1' })
      );

      expect(rootNode.render(getCanvasMock()).canvas).toEqual(['CustomText']);

      expect(customRender).toHaveBeenCalled();
    });
  });

  describe('styling', () => {
    function getTextMock(textNodeProps, text) {
      const container = (({}: any): ContainerNode);
      const rootNode = new ChunkNode(
        container,
        getNodeProps({
          stylizeArgs: textNodeProps,
        })
      );
      rootNode.appendInitialChild(new TextNode(container, { children: text }));

      return rootNode.render(getCanvasMock()).canvas.join('\n');
    }

    it('respects color and background color', () => {
      const colored = getTextMock(
        {
          color: colors.magenta,
          backgroundColor: colors.cyan,
        },
        'Hello world'
      );
      expect(colored).toMatchSnapshot();
    });

    it('respects fontWeight', () => {
      const bold = getTextMock(
        {
          fontWeight: 'bold',
        },
        'Hello world'
      );
      const normal = getTextMock(
        {
          fontWeight: 'normal',
        },
        'Hello world'
      );

      expect(bold).toMatchSnapshot();
      expect(normal).toMatchSnapshot();
    });

    it('respects fontStyle', () => {
      const italic = getTextMock({ fontStyle: 'italic' }, 'Hello world');
      const normal = getTextMock({ fontStyle: 'normal' }, 'Hello world');
      expect(italic).toMatchSnapshot();
      expect(normal).toMatchSnapshot();
    });

    it('respects textDecoration', () => {
      const underline = getTextMock(
        { textDecoration: 'underline' },
        'Hello world'
      );
      const lineThrough = getTextMock(
        { textDecoration: 'line-through' },
        'Hello world'
      );
      const normal = getTextMock({ textDecoration: 'normal' }, 'Hello world');
      expect(underline).toMatchSnapshot();
      expect(lineThrough).toMatchSnapshot();
      expect(normal).toMatchSnapshot();
    });

    it('respects textTransform', () => {
      const capitalize = getTextMock(
        { textTransform: 'capitalize' },
        'Hello world'
      );
      const uppercase = getTextMock(
        { textTransform: 'uppercase' },
        'Hello world'
      );
      const lowercase = getTextMock(
        { textTransform: 'lowercase' },
        'Hello world'
      );
      const none = getTextMock({ textTransform: 'none' }, 'Hello world');
      expect(capitalize).toMatchSnapshot();
      expect(uppercase).toMatchSnapshot();
      expect(lowercase).toMatchSnapshot();
      expect(none).toMatchSnapshot();
    });

    it('respects visibility', () => {
      const visible = getTextMock({ visibility: 'visible' }, 'Hello world');
      const hidden = getTextMock({ visibility: 'hidden' }, 'Hello world');
      expect(visible).toMatchSnapshot();
      expect(hidden).toMatchSnapshot();
    });
  });
});

//   xit('should render tree of nodes to a string', () => {
//     const elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//     };

//     const rootNode = getTreeFixture(container);
//     expect(rootNode.render()).toEqual([
//       '  ',
//       '  ',
//       '  ',
//       '  Text1',
//       '    Text2',
//       '  Text3',
//       '  Text4  ',
//       '  Text5  ',
//       '  Text6',
//     ]);
//   });

//   xit('should memoize rendered elements', () => {
//     const elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//     };

//     const rootNode = getTreeFixture(container);
//     rootNode.render();

//     function runAssertionsForNodes(nodes) {
//       nodes.forEach(node => {
//         if (node instanceof ChunkNode) {
//           expect(node.hasChildrenChanged).toBeFalsy();
//           expect(node.memoizedElements.length).toBeGreaterThan(0);
//           runAssertionsForNodes(node.children);
//         }
//       });
//     }

//     runAssertionsForNodes([rootNode]);

//     expect(rootNode.memoizedElements).toEqual(elements);
//   });

//   xit('should invalidate changed path', () => {
//     let elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//       invalidateParent: jest.fn(),
//     };

//     const rootNode = getTreeFixture(container);
//     rootNode.parent = container;
//     rootNode.render();

//     rootNode.children[0].children[1].children[0].replaceChildren('NewText');

//     expect(container.invalidateParent).toHaveBeenCalled();
//     expect(rootNode.hasChildrenChanged).toBeTruthy();
//     expect(rootNode.children[0].hasChildrenChanged).toBeTruthy();
//     expect(rootNode.children[0].children[1].hasChildrenChanged).toBeTruthy();

//     const oldElements = [...elements];
//     elements = [];
//     rootNode.render();

//     expect(oldElements).not.toEqual(elements);
//     expect(oldElements.length).toBe(elements.length);
//     expect(elements[1].text).toEqual('NewText');
//     expect([elements[0], ...elements.slice(2)]).toEqual([
//       oldElements[0],
//       ...oldElements.slice(2),
//     ]);
//   });

//   xit('should invalidate when new node is added', () => {
//     let elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//       invalidateParent: jest.fn(),
//     };

//     const rootNode = getTreeFixture(container);
//     rootNode.parent = container;
//     rootNode.render();

//     rootNode.children[1].appendChild(
//       // $FlowFixMe
//       new TextNode(container, { children: 'Added' })
//     );

//     expect(rootNode.hasChildrenChanged).toBeTruthy();
//     expect(rootNode.children[0].hasChildrenChanged).toBeFalsy();
//     expect(rootNode.children[1].hasChildrenChanged).toBeTruthy();

//     const oldElements = [...elements];
//     elements = [];
//     rootNode.render();

//     expect(elements).toEqual([
//       ...oldElements.slice(0, 4),
//       elements[4],
//       oldElements[4],
//     ]);
//   });

//   xit('should invalidate when node is removed', () => {
//     let elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//       invalidateParent: jest.fn(),
//     };

//     const rootNode = getTreeFixture(container);
//     rootNode.parent = container;
//     rootNode.render();

//     rootNode.removeChild(rootNode.children[1]);

//     expect(rootNode.hasChildrenChanged).toBeTruthy();

//     const oldElements = [...elements];
//     elements = [];
//     rootNode.render();

//     expect(elements).toEqual([...oldElements.slice(0, 3), oldElements[4]]);
//   });

//   xit('should invalidate children when position is updated', () => {
//     let elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//       invalidateParent: jest.fn(),
//     };

//     const rootNode = getTreeFixture(container);
//     rootNode.parent = container;
//     rootNode.render();

//     rootNode.props.x += 2;
//     rootNode.props.y += 2;

//     const oldElements = [...elements];
//     elements = [];
//     rootNode.render();

//     expect(elements).toEqual(
//       oldElements.map(e => ({
//         ...e,
//         parentsOffsetX: e.parentsOffsetX + 2,
//         parentsOffsetY: e.parentsOffsetY + 2,
//       }))
//     );
//   });

//   xit('should prepend child', () => {
//     let elements = [];
//     const container = {
//       appendElement(element) {
//         elements.push(element);
//       },
//       invalidateParent: jest.fn(),
//     };

//     const rootNode = getTreeFixture(container);
//     rootNode.parent = container;
//     rootNode.render();

//     rootNode.prependChild(
//       // $FlowFixMe
//       new TextNode(container, { children: 'Appended' }),
//       rootNode.children[1]
//     );

//     expect(rootNode.hasChildrenChanged).toBeTruthy();

//     const oldElements = [...elements];
//     elements = [];
//     rootNode.render();

//     expect(elements).toEqual([
//       ...oldElements.slice(0, 3),
//       {
//         x: 0,
//         y: 0,
//         parentsOffsetX: 2,
//         parentsOffsetY: 0,
//         text: 'Appended',
//       },
//       ...oldElements.slice(3),
//     ]);
//   });
// });
