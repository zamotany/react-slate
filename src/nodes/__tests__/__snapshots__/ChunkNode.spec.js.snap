// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodes/ChunkNode styling adds border 1`] = `
"┌───────────┐
│Hello world│
└───────────┘"
`;

exports[`nodes/ChunkNode styling adds border 2`] = `
"╔═══════════╗
║Hello world║
╚═══════════╝"
`;

exports[`nodes/ChunkNode styling adds border 3`] = `"Hello world"`;

exports[`nodes/ChunkNode styling adds border 4`] = `
"[31m┌───────────┐[39m
[31m│[39mHello world[31m│[39m
[31m└───────────┘[39m"
`;

exports[`nodes/ChunkNode styling adds border 5`] = `
"[31m[44m┌───────────┐[49m[39m
[44m[31m│[39mHello world[31m│[39m[49m
[31m[44m└───────────┘[49m[39m"
`;

exports[`nodes/ChunkNode styling adds border 6`] = `
"┌─────────────┐
│Hello world  │
└─────────────┘"
`;

exports[`nodes/ChunkNode styling adds border 7`] = `
"┌─────────────┐
│Hello world  │
│             │
└─────────────┘"
`;

exports[`nodes/ChunkNode styling adds border 8`] = `
"┌─────────────┐
│Hello        │
└─────────────┘"
`;

exports[`nodes/ChunkNode styling aligns text 1`] = `"Hello world"`;

exports[`nodes/ChunkNode styling aligns text 2`] = `"Hello world"`;

exports[`nodes/ChunkNode styling aligns text 3`] = `"Hello world"`;

exports[`nodes/ChunkNode styling aligns text 4`] = `"Hello world          "`;

exports[`nodes/ChunkNode styling aligns text 5`] = `"     Hello world     "`;

exports[`nodes/ChunkNode styling aligns text 6`] = `"          Hello world"`;

exports[`nodes/ChunkNode styling aligns text 7`] = `"Hello world         "`;

exports[`nodes/ChunkNode styling aligns text 8`] = `"     Hello world    "`;

exports[`nodes/ChunkNode styling aligns text 9`] = `"         Hello world"`;

exports[`nodes/ChunkNode styling aligns text 10`] = `
"┌───────────────────┐
│Hello world        │
└───────────────────┘"
`;

exports[`nodes/ChunkNode styling aligns text 11`] = `
"┌───────────────────┐
│    Hello world    │
└───────────────────┘"
`;

exports[`nodes/ChunkNode styling aligns text 12`] = `
"┌───────────────────┐
│        Hello world│
└───────────────────┘"
`;

exports[`nodes/ChunkNode styling respects color and background color 1`] = `"[35m[46mHello world[49m[39m"`;

exports[`nodes/ChunkNode styling respects fontStyle 1`] = `"[3mHello world[23m"`;

exports[`nodes/ChunkNode styling respects fontStyle 2`] = `"Hello world"`;

exports[`nodes/ChunkNode styling respects fontWeight 1`] = `"[1mHello world[22m"`;

exports[`nodes/ChunkNode styling respects fontWeight 2`] = `"Hello world"`;

exports[`nodes/ChunkNode styling respects textDecoration 1`] = `"[4mHello world[24m"`;

exports[`nodes/ChunkNode styling respects textDecoration 2`] = `"[9mHello world[29m"`;

exports[`nodes/ChunkNode styling respects textDecoration 3`] = `"Hello world"`;

exports[`nodes/ChunkNode styling respects textTransform 1`] = `"Hello World"`;

exports[`nodes/ChunkNode styling respects textTransform 2`] = `"HELLO WORLD"`;

exports[`nodes/ChunkNode styling respects textTransform 3`] = `"hello world"`;

exports[`nodes/ChunkNode styling respects textTransform 4`] = `"Hello world"`;

exports[`nodes/ChunkNode styling respects visibility 1`] = `"Hello world"`;

exports[`nodes/ChunkNode styling respects visibility 2`] = `"[8mHello world[28m"`;
